Allow skipping on noarch package from the command line

jjhelmus opened the pull request:
* The building on noarch packages can be controlled from the command line using the `--skip-noarch` argument.
* Only a single build will be done for `noarch: python` recipes.
* Add a `--dry-run` argument to allow builds to be viewed without execution.

jjhelmus committed:
Add --dry-run argument

The --dry-run argument can be used to list what distributions would be built
without actually performing the builds.

jjhelmus committed:
Add a single case for noarch python recipes

jjhelmus committed:
Add --skip-noarch argument to command line

The --skip-noarch argument is used to skip building of noarch recipes.
This option is helpful as these packages only need to be build on a single
platform.  Ons other platforms using this argument can be used to skip these
builds.

jjhelmus commented:
The idea of for the `--skip-noarch` argument is that it would be set on two of the CI services (Travis CI and AppVeyor?) in `staged-recipes`.  Then `noarch` and `noarch: python` python recipes would only be built once.

jjhelmus committed:
Add noarch attribute to DummyPackage

jakirkham commented:
Would it make sense to breakout the `--dry-run` case as one PR and then approach the `noarch` case in a follow-up PR? It seems clear the purpose of `--dry-run` and it seems easy to test and get in. While the `noarch` case definitely has value, we should discuss the approach we take there a bit more.

jakirkham commented:
FTR I'm completely ok with `conda-build-all` learning about `noarch` builds and updating the matrix accordingly. Though I'm debating the merits of actually trying to skip `noarch` builds on the `staged-recipe` CIs at the `conda-build-all` level.

It feels like there is some conflict between `conda-build-all` as a general purpose build tool and as a tool that we use to build at `staged-recipes`. Adding features like `--skip-noarch` don't seem particularly valuable for `conda-build-all` as a general purpose build tool. This might be ok if it really solved the CI usage issue. However it doesn't really solve the CI resource problem either. Meaning that a substantial amount of time will still be spent with CI setup before realizing the build is unnecessary.

This is not to say that we shouldn't solve the CI resource problem. In fact I totally agree that we should. Just this change is raising the question as to whether we can solve the CI resource problem effectively with `conda-build-all` without also impacting its ability to be a general purpose build tool. IMHO we can't solve this problem in `conda-build-all`, at least not effectively, and it may detract from `conda-build-all`'s purpose.

That said, we have had similar discussions about saving CI resources when they are unneeded for a staged-recipe PR (e.g. builds only on specific platforms and now also `noarch`). One relevant thread starts on this [comment]( https://github.com/conda-forge/conda-forge-webservices/issues/79#issuecomment-260768921 ). From some experiments that @goanpeca has done along the lines of canceling unneeded builds, we came to the conclusion that one would need a service that filters out GitHub webhooks from triggering CIs. Whatever this service would look like, I imagine it could take on this CI waste issue created by `noarch` builds at staged-recipes.

This is all to say that if `conda-build-all` already learns about `noarch`, this is helpful both locally and in conda-forge. Also this already saves some build time for those willing to opt-in. Though trying to solve any more than that with `conda-build-all` will not really deliver satisfactory results.

jjhelmus committed:
Revert "Add noarch attribute to DummyPackage"

This reverts commit 9aed8acaababae6a788bd8df28be3aefe0ec5cf0.

jjhelmus committed:
Check for noarch python in conda 4.2 compat manner

Check for noarch python recipes in a manner which fails gracefully on older
versions of conda where the metadata class does not have a 'noarch' attribute.

jjhelmus committed:
Add unit test for noarch: python package

jjhelmus commented:
@jakirkham I agree that the `--dry-run` option as well as having `conda-build-all` produce a single matrix entry for `noarch: python` recipes can and should be their own pull requests.  I'll break them out into separate PRs shortly.

I think the `--skip-noarch` option has some uses outside of conda-forge, although I agree it is less valuable addition than the other changes. On systems with limited resources, for example on a Raspberry Pi Zero, re-using noarch packages built or a more powerful systems can be very helpful.  In these cases `--skip-noarch` could be used to avoid building these packages, although adding the channel where these are available to the inspected channels would accomplish similar results.  

If folks would like to keep `conda-build-all` more general purpose I will not feel bad if `--skip-noarch` is left out.  Mainly, I wanted to offer a possible solution for `staged-recipes` to spark some discussion.

jjhelmus commented:
I think `--skip-noarch` could break non-noarch recipes that depend on a noarch package. Relying on filtering of these recipes from an existing channel is likely a better solution to the problem.

jakirkham commented:
I see. Yeah if it does help in real world build cases, definitely will to reconsider. Though I agree with your analysis that it is possible to use another channel or potentially a prepopulated index to also solve this.

jakirkham commented:
Do you still have interest in this @jjhelmus?

jakirkham commented:
If it is still important to pursue, could you please migrate the PR to ( https://github.com/conda-tools/conda-build-all )?

pelson commented:
For the sake of tidiness, I'll close this PR up. https://github.com/conda-tools/conda-build-all is the new home 👍
